README.md
## About Rockwall Analytics
*Rockwall Analytics* is a values-based organization with a powerful sense of shared purpose towards our mission: to serve people. The brand is to liberate from following a trend, find our own purpose and define our own style, and make that our image

[![SC2 Video](img_gallery/SC2_youtube.png)](https://youtu.be/JVvfk56hckY "SC2 Rockwall Analytics - Click to Watch!")

Our mission is to enable organizations to make an easier transition into the field of Analytics and Applied Ai. Our aim to achieve this more efficiently than an organization might expect were it to invest in the manpower and technology to build such complex platforms. For a universal need such as Data Science and Artificial Intelligence, a dedicated focus is absolutely critical and Rockwall Analytics empowers our clients with the knowledge that allows them to apply in their core business offerings

Data Analytics and Artificial Intelligence are growing to be a ubiquitous need in the modern enterprise ecosystem. Our need for analytics is ranging from basic descriptive and diagnostic analytics to advanced predictive, prescriptive, and cognitive analytics. However, the barrier of entry is high due to expensive infrastructure and highly skilled resource requirements. Rockwall Analytics aims to eliminate this barrier by introducing products that make it easy for businesses to embrace data science and gain data intelligence

> Applying data science technologies like Ai, NLP, and machine learning algorithms can help firms in several areas like fraud detection, risk management, customer sentiment analysis, and personalized marketing

> Data science is disrupting all sector like never before. organizations are sitting on piles of data and harnessing the volumes of data is helping them in various ways, from process automation, process improvements to exploring new delivery models and introducing new services

## DATA & ANALYTICS ADVISORY
It was only in the last few years that *Big Data* became the buzz phrase that confused many and overwhelmed many more. Two years ago, most executives and IT departments were still struggling to understand what the data revolution meant for them and their business; many were just fighting to keep their heads above water

**Rockwall Analytics** understands how to turn data into insights into value, giving your organization the competitive advantage

We combine deep business experience with cutting-edge capabilities and proven solutions to help our clients make better, faster decisions in all areas of their business. In doing so, we can help organizations go beyond their data to increase revenue, reduce costs and manage risk throughout the enterprise

## HOW CAN WE HELP
With analytical and technical solutions deeply anchored in industry and functional knowledge, combined with a detailed understanding of our clients' needs, our approaches demystify *big data*, tackle industry-specific challenges and cover a range of important business issues

*These include*
```
Financial Services
Technology, Media & Telecommunications
Consumer Markets
Energy & Natural Resources
Healthcare
```

![](https://drive.google.com/uc?export=view&id=1i7fzIUxz-oEs8V4uMdoZCQUl51NMrbVz)


basic_calculator/CHANGELOG.txt
Change Log
==========

0.0.1 (25/09/2020)
-------------------
- First Release


Video_Surveillance/test.py
import cv2
import numpy as np
from keras.models import load_model
import argparse
from PIL import Image
import imutils


def mean_squared_loss(x1,x2):
    difference = x1-x2
    a,b,c,d,e = difference.shape
    n_samples = a*b*c*d*e
    sq_difference = difference**2
    Sum = sq_difference.sum()
    distance = np.sqrt(Sum)
    mean_distance = distance/n_samples

    return mean_distance

model = load_model('saved_model.h5')

cap = cv2.VideoCapture('test_video.mp4')
print(cap.isOpened())

while cap.isOpened():
    imagedump = []
    ret,frame = cap.read()


    for i in range(10):
        ret,frame = cap.read()
        image = imutils.resize(frame, width=1000, height=1200)

        frame = cv2.resize(frame, (227,227), interpolation = cv2.INTER_AREA)
        gray = 0.2989*frame[:,:,0] + 0.5870*frame[:,:,1] + 0.1140*frame[:,:,2]
        gray = (gray-gray.mean())/gray.std()
        gray = np.clip(gray, 0, 1)
        imagedump.append(gray)

    imagedump = np.array(imagedump)

    imagedump.resize(227,227,10)
    imagedump = np.expand_dims(imagedump, axis=0)
    imagedump = np.expand_dims(imagedump, axis=4)

    output = model.predict(imagedump)

    loss = mean_squared_loss(imagedump, output)

    if frame.any()==None:
        print('none')


    if cv2.waitKey(10) & 0xFF==ord('q'):
        break
    if loss>0.00068:
        print('Abnormal Event Detected')
        cv2.putText(image, 'Abnormal Event', (220,100), cv2.FONT_HERSHEY_SIMPLEX, 2, (0,0,255), 4)

    cv2.imshow('video',image)

cap.release()
cv2.destroyAllWindows()


Cars/res-net.py
from keras.layers import Input, Add, Dense, Activation, ZeroPadding2D, BatchNormalization, Flatten, Conv2D, AveragePooling2D, MaxPooling2D, GlobalMaxPooling2D
from keras.initializers import glorot_uniform
from keras.models import Model, load_model
from custom_layers.scale_layer import Scale
from keras.optimizers import SGD

# Structure of an identity block:
from identity_block import *

# Structure of a Convolution Block:
from conv_block import *

def ResNet(input_shape = (224, 224, 3), classes = 196):
    """
    Implementation of the popular ResNet50 the following architecture:
    CONV2D -> BATCHNORM -> RELU -> MAXPOOL -> CONVBLOCK -> IDBLOCK*2 -> CONVBLOCK -> IDBLOCK*3
    -> CONVBLOCK -> IDBLOCK*5 -> CONVBLOCK -> IDBLOCK*2 -> AVGPOOL -> TOPLAYER

    Arguments:
    input_shape -- shape of the images of the dataset
    classes -- integer, number of classes

    Returns:
    model -- a Model() instance in Keras
    """
    # Set epsilon value for Batch Normalization
    eps = 1.1e-5

    # Define the input as a tensor with shape input_shape
    X_input = Input(input_shape)


    # Zero-Padding
    X = ZeroPadding2D((3, 3))(X_input)

    # Stage 1
    X = Conv2D(64, (7, 7), strides = (2, 2), name = 'conv1', use_bias=False, kernel_initializer = glorot_uniform(seed=0))(X)
    X = BatchNormalization(epsilon = eps, axis = 3, name = 'bn_conv1')(X)
    X = Scale(axis = 3, name = 'scale_conv1')(X)
    X = Activation('relu')(X)
    X = MaxPooling2D((3, 3), strides=(2, 2))(X)

    # Stage 2
    X = convolutional_block(X, f = 3, filters = [64, 64, 256], stage = 2, block='a', s = 1)
    X = identity_block(X, 3, [64, 64, 256], stage=2, block='b')
    X = identity_block(X, 3, [64, 64, 256], stage=2, block='c')

    # Stage 3 (Default Identity blocks = 3)
    X = convolutional_block(X, f = 3, filters = [128,128,512], stage = 3, block='a', s = 2)
    for i in range(1,8):
      X = identity_block(X, 3, [128, 128, 512], stage=3, block='b'+str(i))

    # Stage 4 (Default Identity blocks = 5)
    X = convolutional_block(X, f = 3, filters = [256, 256, 1024], stage = 4, block='a', s = 2)
    for i in range(1,36):
      X = identity_block(X, 3, [256, 256, 1024], stage=4, block='b'+str(i))


    # Stage 5
    X = convolutional_block(X, f = 3, filters = [512, 512, 2048], stage = 5, block='a', s = 2)
    X = identity_block(X, 3, [512, 512, 2048], stage=5, block='b')
    X = identity_block(X, 3, [512, 512, 2048], stage=5, block='c')

    # First output layer
    Xfc = AveragePooling2D(pool_size=(2, 2), padding='valid', name = "avg_pool")(X)
    Xfc = Flatten()(Xfc)
    Xfc = Dense(1000, activation='softmax', name='fc1000' + str(classes))(Xfc)

    # Transfer learning from pretrained weights
    model = Model(inputs = X_input, outputs = Xfc, name='Cifar10_transfer_learning')
    model.load_weights('models/resnet152_weights_tf.h5', by_name=True)

    # Switch to the final output layer
    Xfc2 = AveragePooling2D((7, 7), name='avg_pool')(X)
    Xfc2 = Flatten()(Xfc2)
    Xfc2 = Dense(classes, activation='softmax', name='fc'+str(classes))(Xfc2)


    # Create model
    model = Model(inputs = X_input, outputs = Xfc2, name='ResNet150')

    return model


Cars/CarModel.py
# coding: utf-8

# Import a bunch of stuff
import numpy as np
from keras import layers, callbacks
from keras.optimizers import SGD, Adam
from keras.layers import Input, Add, Dense, Activation, ZeroPadding2D, BatchNormalization, Flatten, Conv2D, AveragePooling2D, MaxPooling2D, GlobalMaxPooling2D
from keras.models import Model, load_model
from keras.preprocessing import image
from keras.utils import layer_utils
from keras.utils.data_utils import get_file
from keras.applications.imagenet_utils import preprocess_input
import pydot
from IPython.display import SVG
from keras.utils.vis_utils import model_to_dot
from keras.utils import plot_model
from resnets_utils import *
from keras.initializers import glorot_uniform
import dask.array as da
import scipy.misc
from matplotlib.pyplot import imshow
# This only works on jupyter notebook:
# get_ipython().magic('matplotlib inline')

# Clear Memory
import gc
gc.collect()

import keras.backend as K
K.set_image_data_format('channels_last')
K.set_learning_phase(1)

#==========================================================================
# PART 1 - Set Some Stuff
#==========================================================================
img_width = 224
img_height = 224
classes = 196
batch_size = 16
epochs = 10 # Be sure to set this based on how long you want to run for!
patience = 50 # For Callbacks
verbose = 1
num_train_samples = 6549
num_valid_samples = 1695 # Cross validation: num_train_samples + num_valid_samples = # of train images
mode = 'sgd' # adam or sgd

#==========================================================================
# PART 2 - Initialize the Model
#==========================================================================

# Import the structure for an identity block, a convolution block, and a full residual CNN (based on both):
from identity_block import *
from conv_block import *
from res_net import *

# Build the model:
model = ResNet(input_shape = (img_width, img_height, 3), classes = classes)
if (mode == 'sgd'): optimizer = SGD(lr=1e-3, decay=1e-6, momentum=0.9, nesterov=True) #This one seems to work better
if (mode == 'adam'): optimizer = Adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=None, decay=0.0, amsgrad=False)
try:
	model.load_weights("weights.best.hdf5")
        model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])
except:
        model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])

#==========================================================================
# PART 3 - Import Data and Set-up Training Parameters (Callbacks, etc.)
#==========================================================================

# Load Dataset (note that test data is loaded separately lower down - this helps with memory)
X_train, Y_train_orig, classes = load_train_dataset()

# It's not necessary to normalize X_train because of the Scale step in the neural network (after BatchNorm)

# Convert training and test labels to one hot matrices
Y_train = convert_to_one_hot(Y_train_orig, 196).T

# If cross validation is desired, then take some of the training data for this purpose
if (num_valid_samples > 0):
	X_valid = X_train[num_train_samples::,:,:,:]
	Y_valid = Y_train[num_train_samples::]
	X_train = X_train[0:num_train_samples,:,:,:]
	Y_train = Y_train[0:num_train_samples]


print ("number of training examples = " + str(X_train.shape[0]))
print ("X_train shape: " + str(X_train.shape))
print ("Y_train shape: " + str(Y_train.shape))
print ("X_valid shape: " + str(X_valid.shape))
print ("Y_valid shape: " + str(Y_valid.shape))

# Data Augmentation and Cross Validation
train_data_gen = image.ImageDataGenerator(rotation_range=20., width_shift_range=0.1, height_shift_range=0.1, zoom_range=0.2, horizontal_flip=True)
train_data_gen.fit(X_train)
train_generator = train_data_gen.flow(X_train, Y_train, batch_size=batch_size)
valid_data_gen = image.ImageDataGenerator()
valid_generator = valid_data_gen.flow(X_valid, Y_valid, batch_size=batch_size)

# Callbacks
checkpoint = callbacks.ModelCheckpoint("weights.best.hdf5", monitor='val_acc', verbose=verbose, save_best_only=False, save_weights_only=False, mode='auto', period=1)
csv_logger = callbacks.CSVLogger('training.log', append=True)
early_stop = callbacks.EarlyStopping('val_acc', patience=patience)
reduce_lr = callbacks.ReduceLROnPlateau('val_acc', factor=0.1, patience=int(patience/4), verbose=1)
callbacks=[csv_logger,checkpoint,early_stop,reduce_lr]

#==========================================================================
# PART 4 - Train the Model
#==========================================================================
model.fit_generator(train_generator, steps_per_epoch=num_train_samples/batch_size, validation_data=valid_generator, validation_steps=num_valid_samples/batch_size, epochs=epochs, callbacks=callbacks, verbose=verbose)

#==========================================================================
# PART 5 - Test the Model
#==========================================================================

# Tidy up memory:
X_train_orig = Y_train_orig = X_train = Y_train = None

# Imoprt stuff
X_test, Y_test_orig, classes = load_test_dataset()
Y_test = convert_to_one_hot(Y_test_orig, 196).T

print ("number of test examples = " + str(X_test.shape[0]))
print ("X_test shape: " + str(X_test.shape))
print ("Y_test shape: " + str(Y_test.shape))

preds = model.evaluate(X_test, Y_test)
print ("Loss = " + str(preds[0]))
print ("Test Accuracy = " + str(preds[1]))


Cars/conv-block.py
from keras.layers import Input, Add, Dense, Activation, ZeroPadding2D, BatchNormalization, Flatten, Conv2D, AveragePooling2D, MaxPooling2D, GlobalMaxPooling2D

from keras.initializers import glorot_uniform
from custom_layers.scale_layer import Scale

def convolutional_block(X, f, filters, stage, block, s = 2):
    """
    Implementation of the convolutional block as defined in Figure 4

    Arguments:
    X -- input tensor of shape (m, n_H_prev, n_W_prev, n_C_prev)
    f -- integer, specifying the shape of the middle CONV's window for the main path
    filters -- python list of integers, defining the number of filters in the CONV layers of the main path
    stage -- integer, used to name the layers, depending on their position in the network
    block -- string/character, used to name the layers, depending on their position in the network
    s -- Integer, specifying the stride to be used

    Returns:
    X -- output of the convolutional block, tensor of shape (n_H, n_W, n_C)
    """

    # defining name basis
    eps = 1.1e-5
    conv_name_base = 'res' + str(stage) + block + '_branch'
    bn_name_base = 'bn' + str(stage) + block + '_branch'
    scale_name_base = 'scale' + str(stage) + block + '_branch'

    # Retrieve Filters
    F1, F2, F3 = filters

    # Save the input value
    X_shortcut = X

    ##### MAIN PATH #####
    # First component of main path
    X = Conv2D(F1, (1, 1), strides = (s,s), padding = 'valid', name = conv_name_base + '2a', use_bias=False, kernel_initializer = glorot_uniform(seed=0))(X)
    X = BatchNormalization(epsilon = eps, axis = 3, name = bn_name_base + '2a')(X)
    X = Scale(axis=3, name = scale_name_base + '2a')(X)
    X = Activation('relu')(X)

    # Second component of main path
    X = Conv2D(F2, (f, f), strides = (1,1), padding = 'same', name = conv_name_base + '2b', use_bias=False, kernel_initializer = glorot_uniform(seed=0))(X)
    X = BatchNormalization(epsilon = eps, axis = 3, name = bn_name_base + '2b')(X)
    X = Scale(axis=3, name = scale_name_base + '2b')(X)
    X = Activation('relu')(X)

    # Third component of main path
    X = Conv2D(F3, (1, 1), strides = (1,1), padding = 'valid', name = conv_name_base + '2c', use_bias=False, kernel_initializer = glorot_uniform(seed=0))(X)
    X = BatchNormalization(epsilon = eps, axis = 3, name = bn_name_base + '2c')(X)
    X = Scale(axis=3, name=scale_name_base + '2c')(X)

    ##### SHORTCUT PATH ####
    X_shortcut = Conv2D(F3, (1, 1), strides = (s,s), padding = 'valid', name = conv_name_base + '1', use_bias=False, kernel_initializer = glorot_uniform(seed=0))(X_shortcut)
    X_shortcut = BatchNormalization(epsilon = eps, axis = 3, name = bn_name_base + '1')(X_shortcut)
    X_shortcut = Scale(axis=3, name=scale_name_base + '1')(X_shortcut)

    # Final step: Add shortcut value to main path, and pass it through a RELU activation
    X = Add()([X,X_shortcut])
    X = Activation('relu')(X)

    return X


Video_Surveillance/train.py
from keras.preprocessing.image import img_to_array,load_img
import numpy as np
import glob
import os

from keras.layers import Conv3D,ConvLSTM2D,Conv3DTranspose
from keras.stae_models import Sequential
from keras.callbacks import stae_modelCheckpoint, EarlyStopping
import imutils

store_image = []
train_path = './train'
fps = 5
trian_videos = os.listdir('train_path')
train_images_path = train_path + '/frames'
os.makedir(train_images_path)

def store_inarray(image_path):
   	image = load_img(image_path)
    image = img_to_array(image)
    image = cv2.resize(image, (227,227), interpolation = cv2.INTER_AREA)
    gray = 0.2989*image[:,:,0] + 0.5870*image[:,:,1] + 0.1140*image[:,:,2]
    store_image.append(gray)

for video in train_videos:
        os.system( 'ffmpeg -i {}/{} -r 1/{}  {}/frames/%03d.jpg'.format(train_path, video, fps, train_path))
        images = os.listdir(train_images_path)
        for image in images:
            image_path = framepath+ '/'+ image
	        store_inarray(image_path)

store_image = np.array(store_image)
a,b,c = store_image.shape
store_image.resize(b,c,a)
store_image = (store_image-store_image.mean())/(store_image.std())
store_image = np.clip(store_image,0,1)
np.save('training.npy',store_image)


#define stae_model

stae_model = Sequential()
stae_model.add(Conv3D(filters=128, kernel_size=(11,11,1), strides=(4,4,1), padding='valid', input_shape=(227,227,10,1), activation='tanh'))
stae_model.add(Conv3D(filters=64, kernel_size=(5,5,1), strides=(2,2,1), padding='valid', activation='tanh'))
stae_model.add(ConvLSTM2D(filters=64, kernel_size=(3,3), strides=1, padding='same', dropout=0.4, recurrent_dropout=0.3, return_sequences=True))
stae_model.add(ConvLSTM2D(filters=32, kernel_size=(3,3), strides=1, padding='same', dropout=0.3, return_sequences=True))
stae_model.add(ConvLSTM2D(filters=64, kernel_size=(3,3), strides=1, return_sequences=True, padding='same',dropout=0.5))
stae_model.add(Conv3DTranspose(filters=128, kernel_size=(5,5,1), strides=(2,2,1),padding='valid', activation='tanh'))
stae_model.add(Conv3DTranspose(filters=1, kernel_size=(11,11,1), strides=(4,4,1),padding='valid', activation='tanh'))

stae_model.compile(optimizer='adam', loss='mean_squared_error', metrics=['accuracy'])


training_data = np.load('training.npy')
frames = training_data.shape[2]
frames = frames-frames%10

training_data = training_data[:,:,:frames]
training_data = training_data.reshape(-1,227,227,10)
training_data = np.expand_dims(training_data, axis=4)
target_data = training_data.copy()

epochs = 5
batch_size = 1

callback_save = ModelCheckpoint('saved_model.h5',
								monitor='mean_squared_error', save_best_only=True)

callback_early_stopping = EarlyStopping(monitor='val_loss', patience=3)


stae_model.fit(training_data,target_data,
          batch_size=batch_size,
          epochs=epochs,
          callbacks=[callback_save,callback_early_stopping]
          )
stae_model.save('saved_model.h5')
